k8s/k8s.go

```diff
@@ -58,235 +58,221 @@ func CmdAddK8s(ctx context.Context, args *skel.CmdArgs, conf types.NetConf, epID
-			result, err = ipAddrsResult(ipAddrs, conf, args, logger)
-			if err != nil {
-				return nil, err
-			}
-			logger.Debugf("IPAM result set to: %+v", result)
+	case ipAddrsNoIpam != "":
+		if !conf.FeatureControl.IPAddrsNoIpam {
+			e := fmt.Errorf("requested feature is not enabled: ip_addrs_no_ipam")
+			logger.Error(e)
+			return nil, e
 		}
 
-		// Configure the endpoint (creating if required).
-		if endpoint == nil {
-			logger.Debug("Initializing new WorkloadEndpoint resource")
-			endpoint = api.NewWorkloadEndpoint()
+		// ipAddrsNoIpam annotation is set so bypass IPAM, and set the IPs manually.
+		overriddenResult, err := overrideIPAMResult(ipAddrsNoIpam, logger)
+		if err != nil {
+			return nil, err
 		}
-		endpoint.Name = epIDs.WEPName
-		endpoint.Namespace = epIDs.Namespace
-		endpoint.Spec.Endpoint = epIDs.Endpoint
-		endpoint.Spec.Node = epIDs.Node
-		endpoint.Spec.Orchestrator = epIDs.Orchestrator
-		endpoint.Spec.Pod = epIDs.Pod
-		endpoint.Labels = labels
-		endpoint.Spec.Ports = ports
-
-		// Set the profileID according to whether Kubernetes policy is required.
-		// If it's not, then just use the network name (which is the normal behavior)
-		// otherwise use one based on the Kubernetes pod's profile(s).
-		if conf.Policy.PolicyType == "k8s" {
-			endpoint.Spec.Profiles = profiles
-		} else {
-			endpoint.Spec.Profiles = []string{conf.Name}
+		logger.Debugf("Bypassing IPAM to set the result to: %+v", overriddenResult)
+
+		// Convert overridden IPAM result into current Result.
+		// This method fill in all the empty fields necessory for CNI output according to spec.
+		result, err = current.NewResultFromResult(overriddenResult)
+		if err != nil {
+			return nil, err
 		}
 
-		// Populate the endpoint with the output from the IPAM plugin.
-		if err = utils.PopulateEndpointNets(endpoint, result); err != nil {
-			// Cleanup IP allocation and return the error.
-			utils.ReleaseIPAllocation(logger, conf.IPAM.Type, args.StdinData)
+		if len(result.IPs) == 0 {
+			return nil, errors.New("failed to build result")
+		}
+
+	case ipAddrs != "":
+		// When ipAddrs annotation is set, we call out to the configured IPAM plugin
+		// requesting the specific IP addresses included in the annotation.
+		result, err = ipAddrsResult(ipAddrs, conf, args, logger)
+		if err != nil {
 			return nil, err
 		}
-		logger.WithField("endpoint", endpoint).Info("Populated endpoint")
+		logger.Debugf("IPAM result set to: %+v", result)
+	}
+
+	// Configure the endpoint (creating if required).
+	if endpoint == nil {
+		logger.Debug("Initializing new WorkloadEndpoint resource")
+		endpoint = api.NewWorkloadEndpoint()
+	}
+	endpoint.Name = epIDs.WEPName
+	endpoint.Namespace = epIDs.Namespace
+	endpoint.Labels = labels
+	endpoint.Spec.Endpoint = epIDs.Endpoint
+	endpoint.Spec.Node = epIDs.Node
+	endpoint.Spec.Orchestrator = epIDs.Orchestrator
+	endpoint.Spec.Pod = epIDs.Pod
+	endpoint.Spec.Ports = ports
+	endpoint.Spec.IPNetworks = []string{}
```

Comment:
what's this for?

Reply:
We need to reset the IPNetworks, since we're assigning a new IP address. We'll populate the nets from the result a few lines below. 

If we didn't do this, the endpoint would keep the old IP from the previous iteration in its spec.

Reply:
I thought so, but was thinking of a case (haven't fully thought of it, just brainstorming with you here) if we "unassign" the IP here and if releasing the IP fails, what happens then